Dependencies/xr/Source/OpenXR/SceneUnderstanding.cpp

```diff
@@ -0,0 +1,497 @@
+    , UpdatedPlanes(updatedPlanes)
+    , RemovedPlanes(removedPlanes)
+{
+}
+
+SceneUnderstanding::SceneUnderstanding::UpdateSceneObjectsArgs::UpdateSceneObjectsArgs(
+    const XrSpace& sceneSpace,
+    const XrSupportedExtensions& extensions,
+    const XrTime displayTime,
+    std::map<XrSceneObjectKeyMSFT, std::shared_ptr<SceneObject>>& sceneObjects,
+    std::vector<XrSceneObjectKeyMSFT>& updatedObjects,
+    std::vector<XrSceneObjectKeyMSFT>& removedObjects)
+    : SceneSpace(sceneSpace)
+    , Extensions(extensions)
+    , DisplayTime(displayTime)
+    , SceneObjects(sceneObjects)
+    , UpdatedObjects(updatedObjects)
+    , RemovedObjects(removedObjects)
+{
+}
+
+struct SceneUnderstanding::SceneUnderstanding::Impl
+{
+public:
+    Impl() {}
+    void Initialize(const InitOptions& options)
+    {
+        const auto& session = options.Session;
+        const auto& extensions = options.Extensions;
+        if (!extensions.SceneUnderstandingSupported)
+        {
+            return;
+        }
+
+        // Default to most recently provided detection boundary information
+        m_boundaryType = options.DetectionBoundaryType;
+        m_sphereRadius = options.SphereRadius;
+        m_boxDimensions = options.BoxDimensions;
+        m_updateInterval = static_cast<XrTime>(NANOSECONDS_IN_SECOND * options.UpdateIntervalInSeconds);
+
+        if (m_initialized)
+        {
+            // This function will be repeatedly called when setting new detection boundaries.
+            return;
+        }
+
+        XrReferenceSpaceCreateInfo spaceCreateInfo{ XR_TYPE_REFERENCE_SPACE_CREATE_INFO };
+        spaceCreateInfo.referenceSpaceType = XR_REFERENCE_SPACE_TYPE_VIEW;
+        spaceCreateInfo.poseInReferenceSpace = xr::math::Pose::Identity();
+        CHECK_XRCMD(xrCreateReferenceSpace(session, &spaceCreateInfo, m_viewSpace.Put()));
+
+        m_sceneObserverHandle.Reset();
+        m_sceneObserverHandle = xr::CreateSceneObserver(extensions, session);
+
+        m_initialized = true;
+    }
+
+    void UpdateFrame(UpdateFrameArgs& args)
+    {
+        Update(args.DisplayTime, args.SceneSpace, args.Extensions);
+        PopulateFrameArguments(args);
+    }
+
+    void UpdateSceneObjects(UpdateSceneObjectsArgs& args)
+    {
+        Update(args.DisplayTime, args.SceneSpace, args.Extensions);
+        PopulateSceneObjectsArgs(args);
+    }
+
+    void Update(const XrTime time, const XrSpace& space, const XrSupportedExtensions& extensions)
+    {
+        if (!m_initialized)
+        {
+            return;
+        }
+
+        if (m_lastFrameUpdate < time)
+        {
+            m_updatedObjects.clear();
+            m_updatedMeshes.clear();
+            m_updatedPlanes.clear();
+            m_removedObjects.clear();
+            m_removedMeshes.clear();
+            m_removedPlanes.clear();
+
+            switch (m_state)
+            {
+            case State::Ready:
+            {
+                HandleReady(time, space, extensions);

```

Comment:
Nit: The formatting on this makes the switch case kind of hard to read.  I would personally tab over everything under the switch once so the outer scopes are not inline with the inner scopes.

 also any reason to have the breaks outside of the scopes?

Reply:
no preference on the breaks/brackets. I'll go ahead and make these changes :)